core/network/src/extra_requests.rs

```diff
@@ -0,0 +1,424 @@
+	/// Prepare network message corresponding to the request.
+	fn into_network_request(&self, request: ExtraRequest<B>) -> Message<B>;
+	/// Accept response.
+	fn import_response(
+		&self,
+		import_queue: &ImportQueue<B>,
+		who: PeerId,
+		request: ExtraRequest<B>,
+		response: Self::Response,
+	);
+	/// Create peer state for peer that is downloading extra data.
+	fn peer_downloading_state(&self, block: B::Hash) -> PeerSyncState<B>;
+}
+
+/// Manages all extra data requests required for sync.
+pub(crate) struct ExtraRequestsAggregator<B: BlockT> {
+	/// Manages justifications requests.
+	justifications: ExtraRequests<B, JustificationsRequestsEssence>,
+	/// Manages finality proof requests.
+	finality_proofs: ExtraRequests<B, FinalityProofRequestsEssence<B>>,
+}
+
+impl<B: BlockT> ExtraRequestsAggregator<B> {
+	pub(crate) fn new() -> Self {
+		ExtraRequestsAggregator {
+			justifications: ExtraRequests::new(JustificationsRequestsEssence),
+			finality_proofs: ExtraRequests::new(FinalityProofRequestsEssence(None)),
+		}
+	}
+
+	pub(crate) fn justifications(&mut self) -> &mut ExtraRequests<B, JustificationsRequestsEssence> {
+		&mut self.justifications
+	}
+
+	pub(crate) fn finality_proofs(&mut self) -> &mut ExtraRequests<B, FinalityProofRequestsEssence<B>> {
+		&mut self.finality_proofs
+	}
+
+	/// Dispatches all possible pending requests to the given peers.
+	pub(crate) fn dispatch(&mut self, peers: &mut HashMap<PeerId, PeerSync<B>>, protocol: &mut Context<B>) {
+		self.justifications.dispatch(peers, protocol);
+		self.finality_proofs.dispatch(peers, protocol);
+	}
+
+	/// Removes any pending extra requests for blocks lower than the
+	/// given best finalized.
+	pub(crate) fn on_block_finalized<F>(
+		&mut self,
+		best_finalized_hash: &B::Hash,
+		best_finalized_number: NumberFor<B>,
+		is_descendent_of: &F,
+	) -> Result<(), fork_tree::Error<ClientError>>
+		where F: Fn(&B::Hash, &B::Hash) -> Result<bool, ClientError>
+	{
+		self.justifications.on_block_finalized(best_finalized_hash, best_finalized_number, is_descendent_of)?;
+		self.finality_proofs.on_block_finalized(best_finalized_hash, best_finalized_number, is_descendent_of)?;
+		Ok(())
+	}
+
+	/// Retry any pending request if a peer disconnected.
+	pub(crate) fn peer_disconnected(&mut self, who: PeerId) {
+		self.justifications.peer_disconnected(&who);
+		self.finality_proofs.peer_disconnected(&who);
+	}
+
+	/// Clear all data.
+	pub(crate) fn clear(&mut self) {
+		self.justifications.clear();
+		self.finality_proofs.clear();
+	}
+}
+
+/// Manages pending block extra data (e.g. justification) requests.
+/// Multiple extras may be requested for competing forks, or for the same branch
+/// at different (increasing) heights. This structure will guarantee that extras
+/// are fetched in-order, and that obsolete changes are pruned (when finalizing a
+/// competing fork).
+pub(crate) struct ExtraRequests<B: BlockT, Essence> {
+	tree: ForkTree<B::Hash, NumberFor<B>, ()>,
```

Comment:
For the light client this means that in practice all finality proofs will always have length 1, since we will request finality proofs in-order as we import change headers.

Reply:
But we do not pause sync until proof is received. I thought that the same case as for full client is possible here: import header `A` + **request** finality proof for it, then import descendant `B` + **request** proof for it. I.e. there could be two simultaneously requested proofs for the same fork. Or am I wrong here?

Reply:
The request for `B` would be queued and  the networking code will only actually send out a request for `B` after we have received a justification for `A`, this ordering is enforced while requesting. We will only emit parallel requests for blocks in different forks. This is required for the full node since justifications must be imported in-order and the justification itself only justifies one block. This is not the case for finality proofs though, it would be fine to request a finality proof for `B` without having one for `A` because the provider should come back with a finality proof with justifications for both `A` and `B`. I think the current approach is fine just noting that this make the finality proofs always have length 1.

Reply:
Right, but there's no code that squashes all same-fork requests into single one. And probably that isn't a good idea, because we want to get finalization proofs ASAP. And if we wanted proof for block #100 && now we want it for block #700 => we need to wait until justification for block >= #700 is emitted before we will receive proof for #100.

Reply:
I agree that we can keep it like this. It would be more efficient to squash requests for the same fork since the light prover and checker allow producing/importing a finality proof that spans multiple change blocks. In the scenario you mentioned, the prover would make a best effort proof, prove everything until #700 since it doesn't have a justification for it. In practice it's not very useful though, because when syncing we will most likely get justifications attached to change blocks, and when we're synced and miss finality I don't expect we'll get too many epochs behind (otherwise we'll get justifications when syncing those finalized blocks).